p4-16/spec/P4-16-spec.mdk

```diff
@@ -5254,6 +5256,42 @@ types (e.g. `structs`) are copied recursively, and all components
 of `header`s are copied, including "validity" bits. Assignment is
 not defined for `extern` values.
 
+### Compound assignment { #sec-compound-assignment }
+
+Compound assignment operators provide a shorter syntax for assigning
+the result of an arithmetic, bitwise or logical operator. The compound
+assignment operator expressions have the form `LHS op RHS`, where `op`
+is one of `+=`, `-=`, `*=`, `/=`, `%=`, `&=`, `|=`, `&&=`, `||=`, `^=`,
+`<<=`, `>>=`, `|+|=`, `|-|=`, and `LHS`, `RHS` are expressions with
+arithmetic types. The following table shows all P4 assignment operators and
+their equivalent.
+
+|----------|-----------------------------------|------------|---------------|
+| Operator | Operator name                     | Example    | Equivalent of |
++----------+:---------------------------------:+:----------:+:-------------:+
+| `=`      | basic assignment                  | `a = b`    | NA            |
+| `+=`     | addition assignment               | `a += b`   | `a = a + b`   |
+| `-=`     | subtraction assignment            | `a -= b`   | `a = a - b`   |
+| `*=`     | multiplication assignment         | `a *= b`   | `a = a * b`   |
+| `/=`     | division assignment               | `a /= b`   | `a = a / b`   |
+| `%=`     | modulo assignment                 | `a %= b`   | `a = a % b`   |
+| `&=`     | bitwise AND assignment            | `a &= b`   | `a = a & b`   |
+| `|=`     | bitwise OR assignment             | `a |= b`   | `a = a | b`   |
+| `&&=`    | logical AND assignment            | `a &&= b`  | `a = a && b`  |
+| `||=`    | logical OR assignment             | `a ||= b`  | `a = a || b`  |
+| `^=`     | bitwise XOR assignment            | `a ^= b`   | `a = a ^ b`   |
+| `<<=`    | bitwise left shift assignment     | `a <<= b`  | `a = a << b`  |
+| `>>=`    | bitwise right shift assignment    | `a >>= b`  | `a = a >> b`  |
+| `|+|=`   | saturating addition assignment    | `a |+|= b` | `a = a |+| b` |
+| `|-|=`   | saturating subtraction assignment | `a |-|= b` | `a = a |-| b` |
+|----------|-----------------------------------|----------- |---------------|
+
+The behavior of every builtin compound assignment expression with the form `E1 op= E2`,
+where `E1` is a modifiable l-value expression and `E2` is an r-value expression,
+is exactly the same as the behavior of the expression `E1 = E1 op E2`, except that
+`E1` is evaluated only once. Unlike normal assignments, compound assignments evaluate
+the left-hand side before evaluating the right-hand side.
```

Comment:
"Unlike normal assignments": Is this true?  Do you have any reference within the spec that defines the evaluation order of `E1` and `E2` in a normal assignment `E1 = E2`?

The sentence "All other expressions are evaluated left-to-right as they appear in the source program." in the section "Expression evaluation order" could be taken to not apply to `E1 = E2`, because that is a statement, not an expression.  If this strict reading is taken, and no other place in the spec mandates an order of evaluation between `E1` and `E2`, then I would agree that the current spec does not mandate this order.

It does seem a little odd to me to define left-to-right evaluation order within expressions, but NOT mandate that `E1` must be evaluated before `E2` in `E1 = E2`.

Reply:
assignments are evaluated left to right in the compiler. the spec should say that.
it is one of the trickiest parts of the compiler, BTW, because the evaluation of the lhs can have side-effects, e.g.
```
void f(inout bit i);
a[f(i)] = a[f(i)];
```

Reply:
I used as a reference the description of augmented assignment statements in Python:
https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements

> Unlike normal assignments, augmented assignments evaluate the left-hand side before evaluating the right-hand side. For example, a[i] += f(x) first looks-up a[i], then it evaluates f(x) and performs the addition, and lastly, it writes the result back to a[i].

Reply:
@rst0git It sounds like from Mihai's comment above that Python should not be used as a reference here.  He believes that even normal assignments should have their left-hand side evaluated first.

Reply:
Thank you Andy and Mihai! I have updated the pull request accordingly.